### PR TITLE
Improve logging by adding a scope and better handling of inner exceptions

### PR DESCRIPTION
**Pull request by fbjerggaard**
This PR adds a scope to all log messages which makes them easier to filter out by SubscriptionId when filtering in a logging system that supports it.

It also fixes some nullreferenced inner exceptions I encountered at times by making all inner exceptions nullable

### CODE REVIEW

File: src/Core/src/Eventuous.Persistence/EventStore/StoreFunctions.cs [hunk 1]
```diff
@@ -30,10 +30,10 @@ CancellationToken              cancellationToken
                 .NoContext();
 
             return result;
-        } catch (Exception e) {
-            throw e.InnerException?.Message.Contains("WrongExpectedVersion") == true
-                ? new OptimisticConcurrencyException(streamName, e)
-                : e;
+        } catch (Exception e) when (e.InnerException is not null && e.InnerException.Message.Contains("WrongExpectedVersion")) {
+            throw new OptimisticConcurrencyException(streamName, e.InnerException);
+        } catch (Exception) {
```
Comment (alexeyzimarev) [inline src/Core/src/Eventuous.Persistence/EventStore/StoreFunctions.cs]: I don't think this branch is necessary

File: src/Core/src/Eventuous.Subscriptions/Logging/Logger.cs [hunk 2]
```diff
@@ -50,7 +50,6 @@ public class LogContext {
         WarnLog        = GetLogger(LogLevel.Warning);
         ErrorLog       = GetLogger(LogLevel.Error);
 
-        InternalLogger? GetLogger(LogLevel logLevel)
-            => Logger.IsEnabled(logLevel) ? new InternalLogger(Logger, logLevel, SubscriptionId) : null;
+        InternalLogger? GetLogger(LogLevel logLevel) => new InternalLogger(Logger, logLevel, SubscriptionId);
```
Comment (alexeyzimarev) [inline src/Core/src/Eventuous.Subscriptions/Logging/Logger.cs]: Why did you remove the check for `IsEnabled`? Without it, the nullable return makes no sense, but, most importantly, the whole log thing will be called even is the given log level isn't enabled. The whole point of the internal logger is _not_ to do anything when low-level logs are produced but logging for that level isn't enabled.

File: src/Core/src/Eventuous.Subscriptions/Logging/InternalLogger.cs [hunk 3]
```diff
@@ -2,16 +2,27 @@
...
-    public void Log(string message, params object[] args) => logger.Log(logLevel, GetMessage(message), args);
+    public void Log(string message, params object[] args) {
+        using (logger.BeginScope(_scope)) {
+            logger.Log(logLevel, GetMessage(message), args);
+        }
+    }
 
-    public void Log(Exception? exception, string message, params object[] args) => logger.Log(logLevel, exception, GetMessage(message), args);
+    public void Log(Exception? exception, string message, params object[] args) {
+        using (logger.BeginScope(_scope)) {
```
Comment (alexeyzimarev) [inline src/Core/src/Eventuous.Subscriptions/Logging/InternalLogger.cs]: Isn't it adding quite a few allocations? What the purpose of the scope here, I don't fully understand. I would understand a scope for handling each message, which starts in the subscription, and ends in the consumer. But, creating and disposing the scope for logging one log message? Is it for adding the subscription id property to the log message rather than just having it as a string? It's totally possible to do at the subscription level as it creates the log context when it starts. I think scopes are explained in the docs as "a logical operation", which is not the case here.

File: src/Core/src/Eventuous.Subscriptions/Logging/Logger.cs (hunk 2 above)
Comment (fbjerggaard) [reply src/Core/src/Eventuous.Subscriptions/Logging/Logger.cs]: It's been a while since I made this change internally, but as I recall it I made it because when changing the loglevel during runtime (possible with fx Seq) the new loglevel was not picked up by the internallogger because it was not enabled when first called. 
I can see that it is not optimal, but I am unsure how else to fix it

File: src/Core/src/Eventuous.Persistence/EventStore/StoreFunctions.cs (hunk 1 above)
Comment (fbjerggaard) [reply src/Core/src/Eventuous.Persistence/EventStore/StoreFunctions.cs]: It was mainly done for readability - it's easily reverted if the other way is preferred

File: src/Core/src/Eventuous.Subscriptions/Logging/InternalLogger.cs (hunk 3 above)
Comment (fbjerggaard) [reply src/Core/src/Eventuous.Subscriptions/Logging/InternalLogger.cs]: You're right - I hadn't thought about putting it earlier in the flow. I'll work on doing that instead of this

File: src/Core/src/Eventuous.Subscriptions/Logging/Logger.cs (hunk 2 above)
Comment (alexeyzimarev) [reply src/Core/src/Eventuous.Subscriptions/Logging/Logger.cs]: Well, the issue is that debug and trace logs are very verbose, and logging only allows avoiding interpolation, but not boxing. All the properties sent to the logger are `object`, so if there's any value type property - it is boxed. Which is exactly what is avoided by not calling the log at all using null for specific log level-based variables. If runtime log level change is a requirement, it should be handled differently somehow, but I would suggest making an issue just for that and create a separate PR to resolve it.

File: src/Core/src/Eventuous.Persistence/EventStore/StoreFunctions.cs (hunk 1 above)
Comment (alexeyzimarev) [reply src/Core/src/Eventuous.Persistence/EventStore/StoreFunctions.cs]: Yeah, my IDE would complain about that useless branch... Would be nice to get rid of it.

File: src/Core/src/Eventuous.Subscriptions/Logging/Logger.cs (hunk 2 above)
Comment (fbjerggaard) [reply src/Core/src/Eventuous.Subscriptions/Logging/Logger.cs]: Agreed - that's the best way to handle it. I'll revert this and create a seperate issue when I get the time to do some debugging as to what exactly happens

File: src/Core/src/Eventuous.Subscriptions/Logging/InternalLogger.cs (hunk 3 above)
Comment (fbjerggaard) [reply src/Core/src/Eventuous.Subscriptions/Logging/InternalLogger.cs]: I have now moved the log scope into the EventSubscription and the DefaultConsumer - as far as I see it that should minimize the allocations and be more in line with the logical operation aspect

File: src/Core/src/Eventuous.Persistence/AggregateStore/AggregateStoreExceptions.cs [hunk 4]
```diff
@@ -4,21 +4,21 @@
...
 
-public class OptimisticConcurrencyException<T>(StreamName streamName, Exception inner) : OptimisticConcurrencyException(typeof(T), streamName, inner)
+public class OptimisticConcurrencyException<T>(StreamName streamName, Exception? inner) : OptimisticConcurrencyException(typeof(T), streamName, inner)
     where T : Aggregate;
 
 public class AggregateNotFoundException : Exception {
-    public AggregateNotFoundException(Type aggregateType, StreamName streamName, Exception inner)
-        : base($"Aggregate {aggregateType.Name} with not found in stream {streamName}. {inner.Message} {inner.InnerException?.Message}", inner) { }
+    public AggregateNotFoundException(Type aggregateType, StreamName streamName, Exception? inner)
+        : base($"Aggregate {aggregateType.Name} with not found in stream {streamName}", inner) { }
```
Comment (bartelink) [inline src/Core/src/Eventuous.Persistence/AggregateStore/AggregateStoreExceptions.cs:19]: this sentence does not make sense, maybe remove 'with', or make it something like 'No events for Aggregate {name} found in {streamName}' ?

Comment (fbjerggaard) [reply src/Core/src/Eventuous.Persistence/AggregateStore/AggregateStoreExceptions.cs:19]: You're right - I didn't catch this while going through this. I might go through the error messages and make sure they are sensible and makes sense sometime during the weekend
